ludwig/automl/automl.py

```diff
@@ -177,13 +183,76 @@ def train_with_config(
     return AutoTrainResults(experiment_analysis)
 
 
-def _model_select(default_configs):
+def _model_select(
+    dataset: Union[str, pd.DataFrame, dd.core.DataFrame, DatasetInfo],
+    default_configs,
+    user_specified_config,
+):
     """
-    Performs model selection based on dataset.
-    Note: Current implementation returns tabnet by default. This will be
-        improved in subsequent iterations
+    Performs model selection based on dataset or user specified model.
+    Note: Current implementation returns tabnet by default. If the
+    percentage of numerical features is >90%, the concat model is used.
     """
-    return default_configs['concat']
+
+    dataset_info = get_dataset_info(dataset) if not isinstance(
+        dataset, DatasetInfo) else dataset
+    fields = dataset_info.fields
+    row_count = dataset_info.row_count
+
+    total_numerical_feats = 0
+
+    for idx, field in enumerate(fields):
+        missing_value_percent = 1 - float(field.nonnull_values) / row_count
+        dtype = infer_type(field, missing_value_percent)
+        if dtype == NUMERICAL:
+            total_numerical_feats += 1
+
+    percent_numerical_feats = total_numerical_feats / len(fields)
+
+    base_config = default_configs["base_config"]
+
+    # tabular dataset heuristics
+    if len(fields) > 3:
+        if percent_numerical_feats > 0.9:
+            base_config = merge_dict(
+                base_config, default_configs["combiner"]["concat"])
+        else:
+            base_config = merge_dict(
+                base_config, default_configs["combiner"]["tabnet"])
+
+        # override combiner heuristic if explicitly provided by user
+        if user_specified_config is not None:
+            if "combiner" in user_specified_config.keys():
+                model_type = user_specified_config["combiner"]["type"]
+                base_config = merge_dict(
+                    base_config, default_configs["combiner"][model_type])
+    else:
+        # text heuristics
+        for input_feature in base_config["input_features"]:
+            # default text encoder is bert
+            # TODO (ASN): add more robust heuristics
+            if input_feature["type"] == "text":
+                input_feature["encoder"] = "bert"
+                base_config = merge_dict(
+                    base_config, default_configs["text"]["bert"])
+
+            # TODO (ASN): add image heuristics
+
+    # override and constrain automl config based on user specified values
+    if user_specified_config is not None:
```

Comment:
I wonder if something like OmegaConf would help simplify the merging process, particularly now that we also have a base_config and combiner config separation in addition to the user_config here.

https://omegaconf.readthedocs.io/en/2.1_branch/usage.html#omegaconf-merge

Reply:
Thanks for the reference to OmegaConf, looks nice!

My understanding of the complexity of this code is that it is removing config from
the hyperopt parameters section that is overlapping with the config_section.key
that is included in the user_config.  I don't see a particular feature of OmegaConf
that can handle this specialized case.  Apologies if I am missing something here.

Reply:
I think it will end up being a refactor of some things. I can take a look in a follow-up. But my goal would be to just be able to eventually reduce the complexity to "merge a bunch of independent configs".